Support basic full text search CONTAINS predicate

rl626 opened the pull request:
### Description

Currently Babelfish does not support full text search queries. This commit implements the most basic form of T-SQL CONTAINS predicate. It supports CONTAINS query on a single column with the query word being a simple term that specifies a match for an exact word or phrase. This is implemented by rewriting the T-SQL CONTAINS predicate into its Postgres full text search counterpart. Concretely, CONTAINS(col_name, 'word') is rewritten into to_tsvector('fts_contains', col_name) @@ to_tsquery('fts_contains', 'word'), and CONTAINS(col_name, '"word1 word2 word3"') is rewritten into to_tsvector('fts_contains', col_name) @@ to_tsquery('fts_contains', 'word1<->word2<->word3'). The query rewriting is performed at the Bison parser level. To make our implementation match the behavior of SQL Server, we introduce a full text configuration 'fts_contains' that makes Postgres use the same stopwords as SQL Server. tsql_contains.stop is the file that contains those stopwords. We import it into Postgres upon Babelfish initialization by changing the makefile. The full text configuration 'fts_contains' is created from this stoplist upon Babelfish startup, which is done by adding the fts_config.sql file.


### Issues Resolved

Task: BABEL-4238

### Test Scenarios Covered ###
We tested CONTAINS on a single column with the query keyword being simple terms as either words or phrases. Phrases are 2 words or 3 words. 



### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

Commit:
Support basic full text search CONTAINS predicate

Currently Babelfish does not support full text search queries. This
commit implements the most basic form of T-SQL CONTAINS predicate. It
supports CONTAINS query on a single column with the query word being a
simple term that specifies a match for an exact word or phrase. This is
implemented by rewriting the T-SQL CONTAINS predicate into its Postgres
full text search counterpart. Concretely, CONTAINS(col_name, 'word') is
rewritten into to_tsvector('fts_contains', col_name) @@
to_tsquery('fts_contains', 'word'), and CONTAINS(col_name, '"word1 word2
word3"') is rewritten into to_tsvector('fts_contains', col_name) @@
to_tsquery('fts_contains', 'word1<->word2<->word3'). The query rewriting
is performed at the Bison parser level. To make our implementation match
the behavior of SQL Server, we introduce a full text configuration
'fts_contains' that makes Postgres use the same stopwords as SQL Server.
tsql_contains.stop is the file that contains those stopwords. We import it
into Postgres upon Babelfish initialization by changing the makefile. The
full text configuration 'fts_contains' is created from this stoplist
upon Babelfish startup, which is done by adding the fts_config.sql file.

Task: BABEL-4238

Signed-off-by: Robin Li <renyuli@amazon.com>

Commit:
add sql script for upgrade test

Commit:
add sql script for upgrade test

Commit:
add sql script for upgrade test

Commit:
add sql script for upgrade test

Commit:
fix for github test

Commit:
move sql script to before reset to fix mvu issue

Commit:
move sql script to before reset to fix mvu issue

Commit:
move sql script to before reset to fix mvu issue

Commit:
move sql script to before reset to fix mvu issue

Commit:
fix expected output for test failure

thephantomthief committed:
Update dotnet tests to use .NET 7.0 (#1598)

Starting 3rd July 2023, all the pre-cached .NET 3.1 versions will be removed from all operating systems. To fix this, we have updated the dotnet sdk version and TFM to use .NET 7.0.

Signed-off-by: Sharu Goel goelshar@amazon.com

Shriramu2002 committed:
Supporting connect timeout server option (#1576)

Description
In this commit, we add support for the connect timeout server option using the sp_serveroption stored procedure. It specifies the time to wait in seconds before returning from a failed attempt to login to a linked server.

The connect timeout server option is implemented using the free_tds library API dbsetlogintime().

Task: BABEL-4274
Signed-off-by: Shriramu A R shriraar@amazon.com

Shriramu2002 committed:
Supporting connect timeout server option (#1576)

Description
In this commit, we add support for the connect timeout server option using the sp_serveroption stored procedure. It specifies the time to wait in seconds before returning from a failed attempt to login to a linked server.

The connect timeout server option is implemented using the free_tds library API dbsetlogintime().

Task: BABEL-4274
Signed-off-by: Shriramu A R shriraar@amazon.com

Commit:
remove the word 'versions' in test files due to bug in sql_validation test that checks for sys.versions

HarshLunagariya committed:
[OSS-ONLY] Fix minor version upgrade github action workflow (#1594)

This commit reverts a unwanted change introduced in minor-version-upgrade.yml github action by commit aaa7127.

Task: BABEL-OSS
Signed-off-by: Harsh Lunagariya lunharsh@amazon.com

Commit:
Revert "remove the word 'versions' in test files due to bug in sql_validation test that checks for sys.versions"

This reverts commit 66fcb3a509982c70f081f9be845ac6b7fac0c300.

Commit:
clean up merge conflict

rl626 committed:
Merge branch 'BABEL_3_X_DEV' into jira-babel-4238

Commit:
remove the word 'versions' in test files due to bug in sql_validation

Sairakan commented on contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-epilogue.y.c:
```diff
@@ -692,6 +692,49 @@ is_json_query(List *name)
 	}
 }
 
+static Node *
```
Please add header comments to all functions, even static functions, that summarize what they do.

rl626 replied:
Done

Sairakan commented on test/JDBC/input/full_text_search/fts-contains-vu-verify.sql:
```diff
@@ -0,0 +1,409 @@
+GO
+
+-- REST NOT SUPPORTED YET
+-- COMMENTED OUT
```
Rather than commenting these out, I would prefer to leave them in, and just have the expected output reflect the current status of the tests. This will ensure that (1) we are erroring gracefully and not crashing and (2) we do not forget to re-enable these tests in the future

rl626 replied:
Done

Sairakan commented on contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--3.2.0--3.3.0.sql:
```diff
@@ -143,5 +143,85 @@ GRANT SELECT ON sys.shipped_objects_not_in_sys TO PUBLIC;
+-- In particular, the string constant '"good old days"' gets rewritten into 'good <-> old <-> days'
+-- This function performs the string rewriting from '"good old days"' to 'good <-> old <-> days'
+
+CREATE OR REPLACE FUNCTION sys.babelfish_fts_contains_rewrite(IN phrase text)
```
Do we have handling for FORMSOF phrases yet? We need to detect such cases, and raise an error if it is not supported yet.

rl626 replied:
Done. We just added handling of unsupported cases.

shashank-yenni committed:
Creating babelfisgpg_unit Extension (#1586)

Description
At present, babelfish encompasses various testing frameworks such as JDBC, Python, .NET, and fuzz tests. Each of these frameworks serves specific purposes and relies on client drivers. However, we lack a framework dedicated to unit testing. This is the motivation behind this to create a unit testing framework for babelfish.

In this commit, we add unit testing framework which enable us to test any piece of our codebase. The babelfishpg_unit directory is added as a new extension within the contrib folder present in babelfish_extensions. It contains the necessary files for the unit testing framework. Here's a brief explanation of each file:

Makefile: This file specifies the build instructions for the babelfishpg_unit extension.
babelfishpg_unit.control: This files contain metadata and control information about the extension.
babelfishpg_unit.h: This header file contains any necessary macros, definitions, or utility functions required for the unit tests.
babelfishpg_unit—1.0.0.sql: This SQL script defines the necessary functions and setup for running the tests. It acts as the entry point for invoking the unit tests. Once it is invoked, it will trigger the execution of all the tests or specific tests.
babelfishpg_unit.c: This file serves as the main entry point for the unit tests. It includes the necessary headers and implements the logic to execute all the defined tests.
test_money.c, ...: These are the individual test files which contains the test cases and assertions specific to a particular area of functionality or component being tested.
log_files: It is a folder which contains log files created after each test run
Issues Resolved
I have developed the framework in such a way that we can run multiple tests by passing multiple parameters. With this framework, all test functions associated with multiple specified parameters can be successfully executed.

Signed-off-by: Yenni Shashank yeshashi@amazon.com

Suthapalli-Ramya-satya-vasavi-srija committed:
sqlcmd hangs for select query with null XML values (#1591)

Previously the sqlcmd hangs for select query which has XML columns. This
commit fixes by sending NBCRowToken when nullMapSize is equal to
simpleRowSize. And for older clients of TDS 7.2 where NbcRowToken
doesn't exist should send 0xffffffffffffffff specially for XML NULL values.

Task: BABEL-4196
Signed-off-by: vasavi suthapalli <svasusri@amazon.com>

thephantomthief committed:
Add instruction to build and configure Babelfish with SSL (#1599)

* Add instruction to build and configure Babelfish with SSL

ODBC Driver 18 for SQL Server by default will encrypt the connection. So
out of the box configuration of Babelfish cannot be connected to from
SQLCMD using this ODBC Driver.

This commit adds instructions on how to build and configure the
babelfish server with SSL enabled.

Signed-off-by: Sharu Goel <goelshar@amazon.com>

Commit:
fix merge conflict

Commit:
add header comments to all functions

rl626 committed:
Fix backward scan flag for parallel queries (#1604)

* Fix backward scan flag for parallel queries

Currently there is a bug in Babelfish that allows the backward scan flag
to be set for parallel queries. However, backward scans cannot be
parallelized. This bug results in assertion failures in the Postgres engine and
leads Babelfish queries to crash. In this commit, we fix the issue by
disallowing backward scan flag to be set when a query is in parallel
mode. This issue occurs during the implementation of full text search in
Babelfish. Fixing the issue unblocks full text search implementation.

Task: BABEL-4261, BABEL-4281
Signed-off-by: Robin Li <renyuli@amazon.com>

Commit:
uncomment unsupported test cases and raise an error for them

Sairakan commented on test/JDBC/input/datatypes/TestXML.txt:
why is this file in the PR? It shouldn't have anything to do with your code change. Please double check against the tip of the source branch that you have properly merged in from main.

Commit:
fix mvu test

rl626 commented:
Closed and moved to pr #1609.
